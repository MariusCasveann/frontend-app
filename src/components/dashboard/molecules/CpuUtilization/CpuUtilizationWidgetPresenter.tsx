import React, { useState, useEffect } from 'react';
import * as chartjs from 'chart.js';
import { Doughnut } from 'react-chartjs-2';
import LoadingOrErrorPresenter from '../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter';
import { nightblue, lightNightblue, grey } from '../../../../config/colors';
import { hasChartData, hasNoChartData } from '../../helpers/checkChartData';
import './CpuUtilization.css';
import './../common/css/CostWidget.css';
import { CardLegend } from '../../../common/molecules/CardLegend/CardLegend';
import { MESSAGE_WHEN_NO_DATA } from '../../../../utils/constants';

interface CpuWidgetPresenterProps {
    loading: boolean;
    error?: string;
    chartData?: chartjs.ChartData;
}

const buildLabel = (tooltipItem: Chart.ChartTooltipItem, chart: Chart.ChartData) => {
    let name, value;
    const labels: string[][] = [['CPU-Dev', 'CPU-Dev not used'], ['CPU-Prod', 'CPU-Prod not used']];

    if (tooltipItem.datasetIndex !== undefined && tooltipItem.index !== undefined) {
        name = labels[tooltipItem.datasetIndex][tooltipItem.index];
        const data =
            chart.datasets && chart.datasets[tooltipItem.datasetIndex] && chart.datasets[tooltipItem.datasetIndex].data;
        value = data && data[tooltipItem.index];
    }

    const label = value && (value as number).toFixed(2).toString();
    return name + ': ' + label + ' %';
};

export default ({ chartData, loading, error }: CpuWidgetPresenterProps) => {
    const [data, setData] = useState(chartData);

    useEffect(() => {
        setData(chartData);
    }, [chartData]);

    const colors = [nightblue, lightNightblue, grey];

    if (loading || error) {
        return <LoadingOrErrorPresenter error={error} loading={loading} />;
    }

    if (hasChartData(data)) {
        return (
            <div className="domain-legend-container">
                {data && (
                    <React.Fragment>
                        {chartData && <CardLegend labels={chartData.labels} colors={colors} />}
                        <div>
                            <Doughnut
                                data={data}
                                legend={{
                                    display: false
                                }}
                                options={{
                                    circumference: Math.PI,
                                    rotation: Math.PI,
                                    responsive: true,
                                    maintainAspectRatio: false,
                                    tooltips: {
                                        callbacks: {
                                            label: buildLabel
                                        }
                                    }
                                }}
                            />
                        </div>
                    </React.Fragment>
                )}
            </div>
        );
    } else if (hasNoChartData(chartData)) {
        return <p>{MESSAGE_WHEN_NO_DATA}</p>;
    }

    return null;
};
