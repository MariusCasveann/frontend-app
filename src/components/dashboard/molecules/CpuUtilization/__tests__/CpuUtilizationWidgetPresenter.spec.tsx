import * as React from 'react';
import renderer from 'react-test-renderer';
jest.mock('react-chartjs-2', () => ({ Doughnut: 'Doughnut' }));
jest.mock('../../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter', () => 'LoadingOrErrorPresenter');
jest.mock('../../../../common/molecules/CardLegend/CardLegend', () => ({ CardLegend: 'CardLegend' }));
import CpuUtilizationWidgetPresenter from '../CpuUtilizationWidgetPresenter';

const mockData = {
    labels: ['Dev', 'Prod', 'Not used'],
    datasets: [
        {
            label: 'Dev',
            data: [0.0618, 9.382],
            backgroundColor: ['#859ec4', '#D8DADA']
        },
        {
            label: 'Prod',
            data: [0.0309, 9.691],
            backgroundColor: ['#859ec4', '#D8DADA']
        }
    ]
};

const mockEmptyData = {
    labels: [],
    datasets: [
        {
            data: [],
            backgroundColor: '#0c3d8a',
            label: 'Dev'
        },
        {
            data: [],
            backgroundColor: '#859ec4',
            label: 'Prod'
        }
    ]
};

describe('CpuUtilizationYearWidgetPresenter', () => {
    describe('when there are chart data', () => {
        it('renders a Semi Doughnut', () => {
            const wrapper = renderer
                .create(<CpuUtilizationWidgetPresenter error={undefined} loading={false} chartData={mockData} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when there are no chart data', () => {
        it('doesnt render a Semi Doughnut', () => {
            const wrapper = renderer
                .create(<CpuUtilizationWidgetPresenter error={undefined} loading={false} chartData={undefined} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is empty', () => {
        it('renders any data for the product', () => {
            const wrapper = renderer
                .create(<CpuUtilizationWidgetPresenter error={undefined} loading={false} chartData={mockEmptyData} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when there is an error', () => {
        it('renders an error message', () => {
            const wrapper = renderer
                .create(<CpuUtilizationWidgetPresenter error={'Error message'} loading={false} chartData={undefined} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when the data is loading', () => {
        it('renders Loading', () => {
            const wrapper = renderer
                .create(<CpuUtilizationWidgetPresenter error={undefined} loading={true} chartData={undefined} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
