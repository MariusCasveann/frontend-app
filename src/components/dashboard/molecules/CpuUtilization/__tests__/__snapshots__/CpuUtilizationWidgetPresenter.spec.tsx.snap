// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CpuUtilizationYearWidgetPresenter when chart data is empty renders any data for the product 1`] = `
<p>
  No data available, please select another time and/or domain/product.
</p>
`;

exports[`CpuUtilizationYearWidgetPresenter when the data is loading renders Loading 1`] = `
<LoadingOrErrorPresenter
  loading={true}
/>
`;

exports[`CpuUtilizationYearWidgetPresenter when there are chart data renders a Semi Doughnut 1`] = `
<div
  className="domain-legend-container"
>
  <CardLegend
    colors={
      Array [
        "#0c3d8a",
        "#859ec4",
        "#D8DADA",
      ]
    }
    labels={
      Array [
        "Dev",
        "Prod",
        "Not used",
      ]
    }
  />
  <div>
    <Doughnut
      data={
        Object {
          "datasets": Array [
            Object {
              "backgroundColor": Array [
                "#859ec4",
                "#D8DADA",
              ],
              "data": Array [
                0.0618,
                9.382,
              ],
              "label": "Dev",
            },
            Object {
              "backgroundColor": Array [
                "#859ec4",
                "#D8DADA",
              ],
              "data": Array [
                0.0309,
                9.691,
              ],
              "label": "Prod",
            },
          ],
          "labels": Array [
            "Dev",
            "Prod",
            "Not used",
          ],
        }
      }
      legend={
        Object {
          "display": false,
        }
      }
      options={
        Object {
          "circumference": 3.141592653589793,
          "maintainAspectRatio": false,
          "responsive": true,
          "rotation": 3.141592653589793,
          "tooltips": Object {
            "callbacks": Object {
              "label": [Function],
            },
          },
        }
      }
    />
  </div>
</div>
`;

exports[`CpuUtilizationYearWidgetPresenter when there are no chart data doesnt render a Semi Doughnut 1`] = `null`;

exports[`CpuUtilizationYearWidgetPresenter when there is an error renders an error message 1`] = `
<LoadingOrErrorPresenter
  error="Error message"
  loading={false}
/>
`;
